File: src/894.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

#include "tree.h"

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
#define RECURSION (0)
#define ITERATIVE (1)
   public:
    unordered_map<int, vector<TreeNode*>> memo;

    Solution() {
#if (RECURSION)
        cout << "Recursion\n\n";
#elif (ITERATIVE)
        cout << "Iterative\n\n";
#endif
    }
    vector<TreeNode*> allPossibleFBT(int n) {
        vector<TreeNode*> retVal;

        if ((n % 2) == 0) {
            return retVal;
        }

#if (RECURSION)
        if (n == 1) {
            retVal.emplace_back(new TreeNode(0));
            return retVal;
        }

        if (memo.find(n) != memo.end()) {
            retVal = memo[n];
            return retVal;
        }

        for (int i = 1; i < n; i += 2) {
            vector<TreeNode*> left = allPossibleFBT(i);
            vector<TreeNode*> right = allPossibleFBT(n - i - 1);
            for (auto l : left) {
                for (auto r : right) {
                    TreeNode* root = new TreeNode(0, l, r);
                    retVal.emplace_back(root);
                }
            }
        }
        memo[n] = retVal;
#elif (ITERATIVE)
        vector<vector<TreeNode*>> dp(n + 1);

        dp[1].push_back(new TreeNode(0));
        for (int count = 3; count <= n; count += 2) {
            for (int i = 1; i < count - 1; i += 2) {
                int j = count - 1 - i;
                for (auto left : dp[i]) {
                    for (auto right : dp[j]) {
                        TreeNode* root = new TreeNode(0, left, right);
                        dp[count].push_back(root);
                    }
                }
            }
        }

        retVal = dp[n];
#endif

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        int n;
    };
    vector<subject> testData{{7}, {3}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: n = 7
     *  Output: [[0,0,0,null,null,0,0,null,null,0,0],[0,0,0,null,null,0,0,0,0],[0,0,0,0,0,0,0],
     *  [0,0,0,0,0,null,null,null,null,0,0],[0,0,0,0,0,null,null,0,0]]
     *
     *  Input: n = 3
     *  Output: [[0,0,0]]
     */

    Solution cSolution;
    vector<TreeNode*> answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: n = " << testData[i].n << "\n";

        answer = cSolution.allPossibleFBT(testData[i].n);
        cout << "Output: [";
        for (long unsigned int j = 0; j < answer.size(); ++j) {
            int height = getTreeHeight(answer[j]);
            cout << ((j == 0) ? "" : ",");
            cout << "[";
            showTreeOutput(answer[j], height);
            cout << "]";
        }
        cout << "]\n";

        cout << "\n";

        // freeTreeNode(answer);
    }

    return EXIT_SUCCESS;
}

File: src/894.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop

from tree import TreeNode


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self) -> None:
        self.method = 2
        if self.method == 1:
            print("Recursion")
        elif self.method == 2:
            print("Iterative")
        print()

    def allPossibleFBT(self, n: int) -> List[Optional[TreeNode]]:
        retVal = []

        if n % 2 == 0:
            return retVal

        if self.method == 1:
            if n == 1:
                return [TreeNode()]

            for i in range(1, n, 2):
                left = self.allPossibleFBT(i)
                right = self.allPossibleFBT(n - i - 1)
                for l in left:
                    for r in right:
                        root = TreeNode(0, l, r)
                        retVal.append(root)
        elif self.method == 2:
            dp = [[] for _ in range(n + 1)]
            dp[1].append(TreeNode(0))
            for count in range(3, n + 1, 2):
                for i in range(1, count - 1, 2):
                    j = count - 1 - i
                    for left in dp[i]:
                        for right in dp[j]:
                            root = TreeNode(0, left, right)
                            dp[count].append(root)
            retVal = dp[n]

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for n in [7, 3]:
            # /* Example
            #  *  Input: n = 7
            #  *  Output: [[0,0,0,null,null,0,0,null,null,0,0],[0,0,0,null,null,0,0,0,0],[0,0,0,0,0,0,0],
            #  *  [0,0,0,0,0,null,null,null,null,0,0],[0,0,0,0,0,null,null,0,0]]
            #  *
            #  *  Input: n = 3
            #  *  Output: [[0,0,0]]
            #  */
            logging.info("Input: n = %s", n)

            retVal = pSolution.allPossibleFBT(n)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
